proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,432 @@
+ sampling operation.
+E.g. `SampleCmpGrad` behaves like `SampleGrad`, except that the comparison is
+ performed on the result,
+ and likewise for `SampleCmpBias` relative to `SampleBias`.
+More specifically, `SampleCmpGrad` uses explicit gradients to compute the LOD
+ rather than implicit ones from other pixels in the current pixel shader quad,
+ and `SampleCmpBias` computes an implicit LOD and then adds a shader-computed
+ bias value to the result.
+
+Two new HLSL overloads for calculating the level of detail will be added:
+
+```c++
+float <TexObject>::CalculateLevelOfDetail(
+      SamplerComparisonState S,
+      float<num> Location);
+float <TexObject>::CalculateLevelOfDetailUnclamped(
+      SamplerComparisonState S,
+      float<num> Location);
+```
+
+Where `TexObject` represents any texture object type except `Texture2DMS*`.
+`CalculateLevelOfDetail` is only available in the pixel and compute shader
+ stages where implicit derivatives can be calculated.
+
+The semantics of the level of detail calculations are identical to the existing
+ `CalculateLevelOfDetail` overloads.
+The only difference with the new overloads is that they calculate the level of
+ detail using comparison sampler objects.
+
+### DXIL Additions
+
+Two new DXIL intrinsics are added:
+
+```llvm
+declare %dx.types.ResRet.f32 @dx.op.sampleCmpGrad.f32(
+    i32,                  ; opcode
+    %dx.types.Handle,     ; texture handle
+    %dx.types.Handle,     ; sampler handle
+    float,                ; coordinate c0
+    float,                ; coordinate c1
+    float,                ; coordinate c2
+    float,                ; coordinate c3
+    i32,                  ; offset o0
+    i32,                  ; offset o1
+    i32,                  ; offset o2
+    float,                ; compare value
+    float,                ; ddx0
+    float,                ; ddx1
+    float,                ; ddx2
+    float,                ; ddy0
+    float,                ; ddy1
+    float,                ; ddy2
+    float)                ; clamp
+```
+
+Valid resource types for the texture handle parameter and the corresponding
+ active coordinates and offsets are the same as for the `sampleGrad` operation
+ except that `TextureCubeArray` resources are invalid.
+Valid active ddx and ddy are the same as offsets.
+The sampler handle parameter must be of type `SamplerComparisonState`.
+
+```llvm
+declare %dx.types.ResRet.f32 @dx.op.sampleCmpBias.f32(
+    i32,                  ; opcode
+    %dx.types.Handle,     ; texture handle
+    %dx.types.Handle,     ; sampler handle
+    float,                ; coordinate c0
+    float,                ; coordinate c1
+    float,                ; coordinate c2
+    float,                ; coordinate c3
+    i32,                  ; offset o0
+    i32,                  ; offset o1
+    i32,                  ; offset o2
+    float,                ; compare value
+    float,                ; bias: in [-16.f,15.99f]
```

Comment:
Is the bias required to be an immediate value or the range [-16.f,15.99f] is only be validated when the bias is immediate value?

Reply:
Here is the origin of the restriction: https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#inst_SAMPLE_B  

" Restrictions:   ...
                2) the range of srcLODBias is ([-16.0f](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#D3D11_MIP_LOD_BIAS_MIN) to [15.99f](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#D3D11_MIP_LOD_BIAS_MAX));
                   values outside of this range will produce undefined results"

This is reflected in the sampleBias documentation in DXIL.rst.